bazaarvoice/emodb
# Updated databus poll to explicitly expose if there are more results

## Github Issue #

None

## What Are We Doing Here?

Recently we encountered an issue where a client wrote a flood up non-mutative updates to Emo.  As a result most databus subscriptions grew to contain millions of raw events.  However, the databus poller automatically filters out any non-mutative updates and puts a fixed time limit on how long it will work before returning whatever events it has to the caller.  This resulted a paradoxical situation where the client's metrics tracking databus subscription size were in the millions while each call to poll returned an empty list of events.

While the extended consequences of this are substantial this pull request is aimed to address the following limited piece of the overall issue:  Most clients, including Emo's own canary, assume that when poll returns an empty list this implies that the subscription's queue is empty.  In response the clients wait a brief period, in canary's case 1 second, before polling again.  However, in this situation this proper behavior actually made a bad situation worse because the queues were continually growing while the clients paused between polls.

This pull request changes the poll call to explicitly return whether the databus queue for a subscription is empty or not.  The caller then can use this explicit information to re-poll immediately or back off as appropriate, rather than inferring emptiness from an empty returned set.  It also updates the Emo canary to use this new information to poll continuously so long as the queue contains more events.

The HTTP calll for poll has been updated to include a new response header, `X-BV-Databus-Empty`, which will have the value "true" or "false" as appropriate.  Because the body of the poll call has not changed it is fully backwards compatible with existing clients.

## How to Test and Verify

1. Check out this PR
2. Create a databus subscription
3. Create a table and numerous documents
4. Poll the subscription with varying limits and validate the "empty" header returns true or false as appropriate.

## Risk

There are two risks with this change:

1. Any existing client which fails when an unexpected header is present will fail.  Arguably, this is unlikely and not good client behavior in the first place.
2. If the "empty" header isn't working correctly this could encourage clients to poll too aggressively or two slowly, depending on which way the defect goes.

### Level 

`Medium`

### Required Testing

`Manual`

## Code Review Checklist


build and run.

we inevitably come back to this code it will only take hours to figure out, not
days.

a victim of rampaging consistency, and should be using this course of action. 
We don't have coding standards out yet for this project, so please make sure to address any feedback regarding STYLE so the codebase remains consistent.
